File: stubs/csharp/src/BeastMode.RmDdd.Examples/Examples/ECommerce/OrderId.cs
using System;
using System.Collections.Generic;
using BeastMode.RmDdd.Domain;
using BeastMode.RmDdd.Utilities;

namespace BeastMode.RmDdd.Examples.ECommerce
{
    /// <summary>
    /// Value object representing an Order identifier.
    /// Demonstrates proper value object implementation in C#.
    /// </summary>
    public class OrderId : ValueObject
    {
        public Guid Value { get; }
        
        public OrderId(Guid value)
        {
            Value = value;
            ValidateOnConstruction();
        }
        
        public OrderId(string value) : this(Guid.Parse(value))
        {
        }
        
        public static OrderId Generate()
        {
            return new OrderId(Guid.NewGuid());
        }
        
        public override ValidationResult Validate()
        {
            var result = new ValidationResult();
            
            if (Value == Guid.Empty)
            {
                result.AddError("Order ID value cannot be empty");
            }
            
            return result;
        }
        
        protected override IEnumerable<object?> GetEqualityComponents()
        {
            yield return Value;
        }
        
        public override string ToString()
        {
            return $"OrderId({Value})";
        }
        
        // Implicit conversion operators for convenience
        public static implicit operator Guid(OrderId orderId) => orderId.Value;
        public static implicit operator OrderId(Guid value) => new(value);
    }
}

File: stubs/java/src/main/java/com/beastmode/rmddd/examples/ecommerce/OrderId.java
package com.beastmode.rmddd.examples.ecommerce;

import com.beastmode.rmddd.domain.ValueObject;
import com.beastmode.rmddd.utilities.ValidationResult;

import java.util.Objects;
import java.util.UUID;

/**
 * Value object representing an Order identifier.
 * Demonstrates proper value object implementation in Java.
 */
public class OrderId extends ValueObject {
    
    private final UUID value;
    
    public OrderId(UUID value) {
        this.value = Objects.requireNonNull(value, "Order ID value cannot be null");
        validateOnConstruction();
    }
    
    public OrderId(String value) {
        this(UUID.fromString(value));
    }
    
    public static OrderId generate() {
        return new OrderId(UUID.randomUUID());
    }
    
    public UUID getValue() {
        return value;
    }
    
    @Override
    public ValidationResult validate() {
        ValidationResult result = new ValidationResult();
        
        if (value == null) {
            result.addError("Order ID value cannot be null");
        }
        
        return result;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        OrderId orderId = (OrderId) o;
        return Objects.equals(value, orderId.value);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
    
    @Override
    public String toString() {
        return "OrderId{" + value + '}';
    }
}